packages/legend-application/src/components/shared/BasicValueSpecificationEditor.tsx

```diff
@@ -170,23 +171,83 @@ const StringPrimitiveInstanceValueEditor = observer(
     className?: string | undefined;
     setValueSpecification: (val: ValueSpecification) => void;
     resetValue: () => void;
+    valueOptions?:
+      | {
+          options: (string | number)[] | undefined;
+          isLoading: boolean;
+          updateOptions: () => void;
+        }
+      | undefined;
   }) => {
-    const { valueSpecification, className, resetValue, setValueSpecification } =
-      props;
+    const {
+      valueSpecification,
+      className,
+      resetValue,
+      setValueSpecification,
+      valueOptions,
+    } = props;
+    const useSelector = Boolean(valueOptions);
     const value = valueSpecification.values[0] as string;
-    const changeValue: React.ChangeEventHandler<HTMLInputElement> = (event) => {
-      instanceValue_changeValue(valueSpecification, event.target.value, 0);
+    const updateValueSpec = (val: string): void => {
+      instanceValue_changeValue(valueSpecification, val, 0);
       setValueSpecification(valueSpecification);
     };
+    const changeInputValue: React.ChangeEventHandler<HTMLInputElement> = (
+      event,
+    ) => {
+      updateValueSpec(event.target.value);
+    };
+    // custom select
+    const selectedValue = { value: value, label: value };
+    const changeValue = (
+      val: null | { value: number | string; label: string },
+    ): void => {
+      const newValue = val === null ? '' : val.value.toString();
+      updateValueSpec(newValue);
+    };
+    const handleInputChange = (
+      inputValue: string,
+      actionChange: InputActionMeta,
+    ): void => {
+      if (actionChange.action === 'input-change') {
+        updateValueSpec(inputValue);
+        valueOptions?.updateOptions();
+      }
+    };
+    const isLoading = valueOptions?.isLoading;
+    const queryOptions = valueOptions?.options?.length
+      ? valueOptions.options.map((e) => ({
+          value: e,
+          label: e.toString(),
+        }))
+      : undefined;
+    const noOptionsMessage =
+      valueOptions?.options === undefined ? (): null => null : undefined;
+
     return (
       <div className={clsx('value-spec-editor', className)}>
-        <input
```

Comment:
I think based on the option provided, we will render this component as `typeahead`-ready or not, so we should keep `input`